fog/sample-paykit/src/cached_tx_data/mod.rs

```diff
@@ -390,6 +460,127 @@ impl CachedTxData {
+        missed_block_indices_sorted.sort_unstable();
+
+        let mut updated_missed_block_ranges: Vec<common::BlockRange> = Vec::new();
+
+        if missed_block_indices_sorted.is_empty() {
+            return updated_missed_block_ranges;
+        }
+
+        if missed_block_indices_sorted.len() == 1 {
+            let first_block_index = missed_block_indices_sorted[0];
+            updated_missed_block_ranges.push(common::BlockRange {
+                start_block: first_block_index,
+                end_block: first_block_index + 1,
+            });
+            return updated_missed_block_ranges;
+        }
+
+        let mut start_block_index = missed_block_indices_sorted[0];
+        let mut i: usize = 0;
+        let mut j: usize = 1;
+        while j < missed_block_indices_sorted.len() {
+            let first_block_index = missed_block_indices_sorted[i];
+            let second_block_index = missed_block_indices_sorted[j];
+            // We've found a gap in indices, so the current range is complete
+            // and should be added.
+            if second_block_index - first_block_index > 1 {
+                let block_range = common::BlockRange {
+                    start_block: start_block_index,
+                    end_block: first_block_index + 1,
+                };
+                updated_missed_block_ranges.push(block_range);
+                start_block_index = second_block_index;
+            }
+
+            i += 1;
+            j += 1;
+        }
+
+        let last_block_index = missed_block_indices_sorted.last().unwrap();
+        let last_block_range = common::BlockRange {
+            start_block: start_block_index,
+            end_block: last_block_index + 1,
+        };
+
+        updated_missed_block_ranges.push(last_block_range);
+
+        updated_missed_block_ranges
+    }
+
+    fn create_missed_block_indices(missed_block_ranges: &[common::BlockRange]) -> HashSet<u64> {
+        let mut missed_block_indices: HashSet<u64> = HashSet::new();
+        for missed_block_range in missed_block_ranges {
+            for missed_block_index in missed_block_range.start_block..missed_block_range.end_block {
+                missed_block_indices.insert(missed_block_index);
+            }
+        }
+
+        missed_block_indices
+    }
+
+    // Converts a ledger::BlockResponses to a Vec<TxOutRecord>.
+    fn create_tx_out_records(block_response: &ledger::BlockResponse) -> Vec<TxOutRecord> {
+        let mut tx_out_records: Vec<TxOutRecord> = Vec::new();
+
+        for block_data in &block_response.blocks {
+            let number_of_tx_outs_in_block = block_data.outputs.len() as u64;
+            let first_tx_out_global_index =
+                block_data.global_txo_count - number_of_tx_outs_in_block;
+
+            for (i, external_tx_out) in block_data.outputs.iter().enumerate() {
+                let tx_out: TxOut = TxOut::try_from(external_tx_out).unwrap();
```

Comment:
Not sure if we made an attempt at proper error handling in other parts of this code, but if so then `unwrap` here is not ideal.

Reply:
yeah it would be better to handle this error by adding it to the error list or something